### PR TITLE
[vlm] adjust max length for special tokens

### LINKED ISSUE #37222: Qwen fails ungracefully when images are truncated
### System Info

- `transformers` version: 4.49.0
- Platform: Linux-6.8.0-1025-gcp-x86_64-with-glibc2.39
- Python version: 3.11.10
- Huggingface_hub version: 0.29.3
- Safetensors version: 0.5.3
- Accelerate version: 0.34.2
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - debug: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- DeepSpeed version: not installed
- PyTorch version (GPU?): 2.6.0+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA H100 80GB HBM3

### Who can help?

@qubvel 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

The following fails when "forward" is called. If you increase the MAX_LENGTH to 30 it succeeds

```
import torch
from transformers import AutoProcessor, Qwen2VLForConditionalGeneration
from PIL import Image

device = "cuda:0"

MAX_LENGTH = 15

# Load model and tokenizer
model_name = "Qwen/Qwen2-VL-7B"
model = Qwen2VLForConditionalGeneration.from_pretrained(
    model_name,
    torch_dtype=torch.float16,
    device_map=None,
)
processor = AutoProcessor.from_pretrained(model_name)

# Prepare model with FSDP
model = model.to(device)

text = "test this image <|vision_start|><|image_pad|><|vision_end|>"
image = [Image.new('RGB', (100, 100), color='red')]

# Prepare inputs
inputs = processor(
    text = text,
    images = image,
    return_tensors="pt",
    max_length=MAX_LENGTH,
    truncation = "longest_first",
    padding = True,
)    
# Move inputs to device
inputs = {k: v.to(device) for k, v in inputs.items()}

inputs["labels"] = inputs["input_ids"].clone()

outputs = model(**inputs)
```

### Expected behavior

I expect the script to either:

1. fail gracefully at tokenization time raising an error informing the user that the image tokens are being truncated and this is untennable, possibly with another kwarg telling it to ignore the error and return the broken tokens or

2. Truncate the image tokens, grid, and pixel values in a compatible way that works with the model forward. This is complicated by the other issue I raised: #37186

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/37222

Instead of raising errors, we adjust the max length manually and log the new length. This way is simpler, than the first idea I had because tokenizer's "overflow_tokens" do not work consistently for slow vs fast tokenizers

TODO: forgot about PG and mistral

For ex if a users sets max length to a small value and sets `truncation=True`, we will crop the input text including image tokens. It doesn't raise any error or warning, up until when inputs are forwarded to model and a mismatch occurs between `placeholder_in_text` and `image_hidden_state.shape`.
The below ex from now on will return all image tokens, even though max length is 20. Also we'll raise warning, so users don't get confused as to why `20` is not the length they got 

```python
       processor = LlavaProcessor.from_pretrained("llava-hf/llava-1.5-7b-hf")

        input_str = self.prepare_text_inputs(batch_size=2)
        image_input = self.prepare_image_inputs(batch_size=2)
        input_str = [f"<image>{sample}" for sample in input_str]
        inputs_truncated = processor(
            text=input_str,
            images=image_input,
            return_tensors="pt",
            truncation=True,
            padding=True,
            max_length=20,
        )


```

### CODE REVIEW

**Review by qubvel — COMMENTED** [review]
Hey @zucchini-nlp, can you please provide a simple example how does it help, cause I'm not sure I understand why increasing `max_length` can help avoid truncating vision tokens.

**Comment by qubvel** [comment]
Thanks, I suppose I got it, the only thing that bothers me - we do not respect `max_length` anymore, and it's becoming a variable. It would also be a breaking change, no? It will increase the max_length for existing pipelines, so maybe better to raise an error

**Comment by zucchini-nlp** [comment]
@qubvel I don't think it's breaking. The length is only adjusted when `truncation=True` with a `max_length` set. Before this PR, users either 1) hit an error during forward pass, or 2) had a high enough `max_length` already. In case (1), we now handle it by adjusting length; in case (2), nothing changes since truncation only affects overly long inputs.

The only potentially breaking case is when `truncation=True, max_length=LEN, padding="max_length"`. We’ll pad to a different `max_length` from now on.

However we can't just raise an error on user-defined `max_length`. Ideally, we'd check if the expanded input exceeds `max_length` using `overflowing_tokens`, but that flag is unreliable and calling the tokenizer twice is inefficient. So I opted for logging a warning instead. A few users padding to `max_length` may need to adjust based on warning message. LMK if you have another way without increasing processing time, I would also love to just raise error and let users handle it themselves. Less hidden code, the better for us tbh

**Comment by qubvel** [comment]
Can we count the number of image tokens IDs for each sample and compare with the expected? smt like `input_ids.isin(image_token_ids).sum(0) < expexted` should be pretty fast I suppose

**Comment by zucchini-nlp** [comment]
Oke, let's merge this after https://github.com/huggingface/transformers/pull/37362, otherwise many tests will be failing. Updated code to raise error

**Comment by zucchini-nlp** [comment]
@qubvel the blocker PR is merged so this can be reviewed now

**Review by qubvel — APPROVED** [review]
Great work @zucchini-nlp! Just let's make sure tests are fixed 👍
